types/src/block_metadata.rs

```diff
@@ -22,54 +20,38 @@ use serde::{Deserialize, Serialize};
 ///    transaction will be executed before all of the user-submitted transactions in the blocks.
 /// 3. Once that special resource is modified, the other user transactions can read the consensus
 ///    info by calling into the read method of that resource, which would thus give users the
-///    information such as the current leader.
+///    information such as who participates in the block.
 #[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub struct BlockMetadata {
-    id: HashValue,
     round: u64,
     timestamp_usecs: u64,
-    // The vector has to be sorted to ensure consistent result among all nodes
-    previous_block_votes: Vec<AccountAddress>,
-    proposer: AccountAddress,
+    // masks on the current validator set
+    participants: Vec<bool>,
+    is_nil: bool,
```

Comment:
Just pointing out that one thing we lose here is the ability to know the proposer from the voter from on-chain state.  This means that externally there isn't visibility into the fairness of proposers in an epoch.  How can the public ensure that it isn't just one validator proposing all the committed blocks?

One option could be to keep the AccountAddress (16 bytes) for the proposer or an index (4 bytes for u32) into the current validator set to retain this information.

Reply:
the index can't handle the nil case well. I can only think of using u32::max to represent a nil. @davidiw pointed out it too, but I don't think we need to expose this information, it's just nice to have.

Reply:
While I appreciate the goal of saving space (network/storage), this current proposal loses information on the proposer, adds more work for the consumer to match against the validator set, and adds the possibility of unexpected edge cases.

Is it worth it?  Given that these txns should be a small percent of the workload once this network is more widely used (e.g. 10s or 100s of tps) and that in the future we don't need the full txn history to sync state, I prefer the existing method.

Network compression can be added to mimic the same improvements as in rocksdb and we can experiment with libstdz and other compression techniques.

A compromise might be to just keep everything in indices of the validator set but as you mention, it's a bit unclean as well.

Reply:
Re the index and null, I’ve seen similar models where the bitmask starts counting from index=1 and the first but always be zero simulating an imaginary “null” participant. Then  leader_index=0 represents the null.